lib/Dialect/Calyx/Export/CalyxEmitter.cpp

```diff
@@ -177,12 +180,56 @@ void Emitter::emitCell(CellOp op) {
   indent() << op.instanceName() << " = " << op.componentName() << "();\n";
 }
 
+void Emitter::emitAssignment(AssignOp op) {
+  // TODO(Calyx): Support guards.
+  if (op.guard())
+    emitOpError(op, "guard not supported for emission currently");
+
+  auto emitAssignmentValue = [&](auto assignValue) {
+    auto definingOp = assignValue.getDefiningOp();
+    std::string emitted;
+    TypeSwitch<Operation *>(definingOp)
+        .Case<CellOp>([&](auto op) {
+          // A cell port should be defined as <instance-name>.<port-name>
+          auto opResult = assignValue.template cast<OpResult>();
+          unsigned portIndex = opResult.getResultNumber();
+          auto ports = getComponentPortInfo(op.getReferencedComponent());
+          StringAttr portName = ports[portIndex].name;
+
+          emitted = op.instanceName();
+          emitted += ".";
+          emitted += portName.getValue();
+        })
+        .template Case<hw::ConstantOp>([&](auto op) {
+          // A constant is defined as <bit-width>'<base><value>, where the base
+          // is `b` (binary), `o` (octal), `h` hexadecimal, or `d` (decimal).
+
+          // Emit the Radix-10 version of the ConstantOp.
+          APInt value = op.value();
+          emitted += std::to_string(value.getBitWidth());
+          emitted += "'d";
+
+          SmallVector<char> stringValue;
+          value.toString(stringValue, /*Radix=*/10, /*Signed=*/false);
+          for (char ch : stringValue)
+            emitted.push_back(ch);
+        })
+        .Default([&](auto op) {
+          emitOpError(op, "not supported for emission inside an assignment");
+        });
+    return emitted;
+  };
+  indent() << emitAssignmentValue(op.dest()) << " = "
```

Comment:
I'm wondering about the use of the same emitter routine for the lhs and rhs. For example, in C, an lvalue is quite different than an rvalue (i.e. in calyx, a constant can never be an "lvalue"). I haven't really thought this through, but would it be helpful to have separate routines for the lhs and the rhs here? That could be useful if those routines can be re-used in other contexts (i.e. the "rvalue" emitter routine may pop up in a few places other than assignments). That could also be a future refactoring...

Reply:
I think a catch-all `emitValue` is fine. Reasoning: I'd like to verify the correctness of {l, r}-values with the Op verifier (i.e. #1520), to avoid doing the same work again here. Does that make sense?